### DISCUSSION TITLE
ParseLiveQuery for older ParseServers

### DISCUSSION THREAD

**Discussion by proj-sashido** [netreconlab/Parse-Swift#102] [Q&A] [answered]
Hi! 
I’m having trouble with getting ParseLiveQuery to work because I’m on an older version of ParseServer (v3.6.0) which is not self-hosted, so I can’t update it beyond this for the time being. 

When I try to connect, I am told the connection status is `.socketNotEstablished`. 
I looked at the Server logs, and I am getting the error: 
```swift
Connect message error Additional properties not allowed
```

I did a bit of googling, and found that it may still be possible to get it working if I can set the property `additionalProperties` to false. 
see: https://bytemeta.vip/repo/parse-community/Parse-SDK-JS/issues/1136

There’s a few references to `additionalProperties` in the Parse-Swift SDK, but most of the time they’re hardcoded to `true`:
```swift
extension LiveQuerySocket {
    func connect(_ task: URLSessionWebSocketTask) async throws {
        let encoded = try ParseCoding.jsonEncoder()
            .encode(await StandardMessage(operation: .connect,
                                          additionalProperties: true))
        // …
    }
}
```

Do you know if setting this to `false` might help with connecting an older ParseServer version to LiveQueries?
Is it possible to have this be a modify-able setting?

EDIT:
I tried changing all the `true` to `false` for `additionalProperties` using a local version of Parse-Swift. Unfortunately I’m getting still getting the error, so I don’t think that’s it :cry:

**Comment by cbaker6** [comment]
> I’m having trouble with getting ParseLiveQuery to work because I’m on an older version of ParseServer (v3.6.0) which is not self-hosted, so I can’t update it beyond this for the time being.

Parse Server [3.6.0](https://github.com/parse-community/parse-server/releases/tag/3.6.0) was released in July 2019 and ParseSwift [1.0.0](https://github.com/parse-community/Parse-Swift/releases/tag/1.0.0) was released in January 2021, so I have never tested LiveQuery with such an old server. In addition, I'm confident a server so old has a ton of security holes...

I don't know of any quick fixes that will allow LiveQuery to work with your server. I recommend looking into moving your server to something that lets you use Parse Server 5+ as I will only add features that work with [Parse LTS servers](https://github.com/parse-community/parse-server#long-term-support)

**Comment by proj-sashido** [comment]
Yeah that’s understandable.
We’re not in a position to change our server hosting just yet, so might just have to use ParseCore.

Thanks for the advice

**Reply by cbaker6** [reply]
You can try out #103, but its possible you will run into additional issues. Note that fixes to support old server will most likely not be added in the future.

**Reply by proj-sashido** [reply]
I am super appreciative of your help, but I don’t think that worked for me. I’m on ParseSwift v5.6.0

Here is the code I used:
```swift
Task {
    do {
        let results = try await query.find()
        print("Results.count", results.count)
                
        let _ = try await query.subscribeCallback()
                
        print("DefaultClient", ParseLiveQuery.client)
        print("DefaultClient.status", ParseLiveQuery.client?.status)
        try await ParseLiveQuery.client?.sendPing()
        print("received pong")
                
        let urlLQ = URL(string: liveQueryURL)
        let PLQ = try await ParseLiveQuery(serverURL: urlLQ)
        print("PLQ", PLQ)
        print("PLQ.status pre-open", PLQ.status)
        PLQ.open { error in print("Open Error", error) }
        print("PLQ.status post-open", PLQ.status)
    }
    catch { print("Thrown Error", error) }
}
```

And the console logs show:
```swift
Results.count 12
DefaultClient Optional(<ParseSwift.ParseLiveQuery: 0x6000024b7cd0>)
DefaultClient.status Optional(ParseSwift.ParseLiveQuery.ConnectionStatus.socketNotEstablished)
received pong
PLQ <ParseSwift.ParseLiveQuery: 0x6000024c4000>
PLQ.status pre-open socketNotEstablished
PLQ.status post-open socketNotEstablished
Open Error Optional(ParseError code=-1 error=ParseLiveQuery Error: attempted to open socket 1 time(s))
```


and my initialisation code:
```swift
let credentials = ParseConfiguration(
    applicationId:  appID,
    clientKey:      clKey,
    serverURL:      url,
    liveQueryServerURL: liveQueryURL,
    usingPostForQuery: true,
    liveQueryConnectionAdditionalProperties: false,
    liveQueryMaxConnectionAttempts: 1
    )
    
let _ = try? await initialize(configuration: credentials)
```

If I’ve missed something silly, please let me know!
I’ll be trying my best to push our host to update the ParseServer, and pushing my organisation to use a better host, failing that.
Hopefully one of those things happens in a timely fashion!

Thank you again for the help! It’s really good to know that if we adopt this, that help will always  be around the corner!

**Reply by cbaker6** [reply]
Your configuration looks correct. You can remove `liveQueryMaxConnectionAttempts: 1` as that was only added for testingI don’t think it will fix your issue though, it seems the older server has a number of other limitations. 

As a note, `usingPostForQuery: true` is saying you don't want to cache queries which is not typical for most developers.

**Comment by cbaker6** [comment, marked as answer]
I've updated the feature comparison table in #72 to mention that ParseSwift should be be used with Parse LiveQuery Server `>=4.0.0`
